ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java

```diff
@@ -365,30 +366,35 @@ private LogicalSchema rebuildWithPseudoAndKeyColsInValue(
 
     int valueIndex = nonPseudoAndKeyCols.size();
 
-    final List<Column> pseudoColumns = new ArrayList<>();
+    final List<Pair<ColumnName, SqlType>> pseudoColumns = new ArrayList<>();
 
     if (pseudoColumnVersion >= ROWTIME_PSEUDOCOLUMN_VERSION) {
-      pseudoColumns.add(Column.of(ROWTIME_NAME, ROWTIME_TYPE, VALUE, 0));
+      pseudoColumns.add(Pair.of(ROWTIME_NAME, ROWTIME_TYPE));
     }
 
     if (pseudoColumnVersion >= ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION) {
-      pseudoColumns.add(Column.of(ROWPARTITION_NAME, ROWPARTITION_TYPE, VALUE, 0));
-      pseudoColumns.add(Column.of(ROWOFFSET_NAME, ROWOFFSET_TYPE, VALUE, 0));
+      pseudoColumns.add(Pair.of(ROWPARTITION_NAME, ROWPARTITION_TYPE));
+      pseudoColumns.add(Pair.of(ROWOFFSET_NAME, ROWOFFSET_TYPE));
     }
 
-    final List<Column> toAddToBuilder = pseudoColumns
-        .stream()
-        .filter(col ->
-            !(SystemColumns.isDisallowedInPullOrScalablePushQueries(col.name(), pseudoColumnVersion)
-                && isPullOrScalablePushQuery)
-        )
-        .collect(Collectors.toList());
-
-    for (Column column : toAddToBuilder) {
-      final Column columnWithIndex = Column.of(column.name(), column.type(), column.namespace(), valueIndex++);
-      builder.add(columnWithIndex);
+    final List<Column> toAddToBuilder = new ArrayList<>();
+
+    //if query is pull or scalable push, need to check if column is disallowed
+    if(isPullOrScalablePushQuery) {
+      for (Pair<ColumnName, SqlType> pair : pseudoColumns) {
+        if(!SystemColumns.isDisallowedInPullOrScalablePushQueries(pair.left, pseudoColumnVersion)) {
+          toAddToBuilder.add(Column.of(pair.left, pair.right, VALUE, valueIndex++));
```

Comment:
Rather than adding columns to `toAddToBuilder` and then immediately adding them to the builder, let's just add to the builder directly here. You can add an "else" branch to the `if(isPullOrScalablePushQuery)` that just adds all the pseudocolumns.